include/range/v3/view/concat.hpp

```diff
@@ -257,8 +257,9 @@ namespace ranges
               : rng_(&rng)
               , its_{emplaced_index<cranges-1>, end(std::get<cranges-1>(rng.rngs_))}
             {}
-            template<bool Other>
-            CPP_ctor(cursor)(cursor<Other> that)(requires IsConst && (!Other))
+            CPP_template(bool Other)(
+                requires IsConst && (!Other))
+            cursor(cursor<Other> that)
```

Comment:
They are ill-formed with no diagnostic required in non-concepts mode; there is no valid specialization of this member template when `IsConst` is `false`, for example. (There's a detailed discussion in the commit message.)

Reply:
There you suggest changing to:

```c++
template<bool Y, bool Z = false,
    enable_if_t<Z || X && Y>* = nullptr>
  S(S<Y>);
```

but you didn't actually define these constructor that way. I see no fundamental difference between what is in this PR and what is currently in range-v3.

Reply:
The difference is hidden inside the preprocessor DSL. In non-concepts mode, this:
```c++
template<bool X>
struct before {
    template<bool Y>
    CPP_ctor(before)(before<Y>)(
        requires X && !Y)
    {}
};

template<bool X>
struct after {
    CPP_template(bool Y)(
        requires X && !Y)
    after(after<Y>)
    {}
};
```
preprocesses to:
```c++
template<bool Const>
struct before {
    template<bool Other>
    before(before<Other>, ::concepts::detail::enable_if_t<
        ::concepts::detail::Nil,
        CPP_false(::concepts::detail::xNil{}) ||
            static_cast<bool>(Const && (!Other))> = {})
    {}
};

template<bool Const>
struct after {
    template<bool Other, bool CPP_false_ = false,
        ::concepts::detail::enable_if_t<int,
            static_cast<bool>(Const && (!Other)) || CPP_false_> = 0>
    after(after<Other>)
    {}
};
```

`before`'s constructor template is ill-formed NDR when `!Const`: `CPP_false(::concepts::detail::xNil{})` is non-dependent, so a compiler could discover that the `enable_if` condition is `false || (false && (!Other))` which is never `true`. `after`'s constructor template is slightly less* IFNDR, since `CPP_false` is dependent and its value could be specified explicitly were this anything other than a constructor template. (In practice, MSVC diagnoses `before` and doesn't diagnose `after`.)

* I'm not terribly enthused by the "slightly less", but there's no good way without concepts to avoid the "no valid specialization" rule for a constructor template that's conditioned on a conjunction of dependent and non-dependent expressions.

Reply:
> there's no good way without concepts to avoid the "no valid specialization" rule for a constructor template that's conditioned on a conjunction of dependent and non-dependent expressions.

For completeness: the effective-but-unattractive way to do so would be to refactor the class template into `Const` / non-`Const` specializations which do / do not have the constructor template:
```c++
template<bool Const>
struct common_behaviors {
    // ...
};

template<bool Const> struct correct;

template<>
struct correct<false> : common_behaviors<false> {
    using common_behaviors<false>::common_behaviors;
    friend correct<true>;
};

template<>
struct correct<true> : common_behaviors<true> {
    using common_behaviors<true>::common_behaviors;
    correct(correct<false>) {}
};
```